plugins/nextcloud-rule-exclusions-before.conf

```diff
@@ -834,15 +866,17 @@ SecRule REQUEST_FILENAME "@endsWith /login/webauthn/finish" \
     t:none,\
     nolog,\
     ver:'nextcloud-rule-exclusions-plugin/1.0.0',\
-    ctl:ruleRemoveTargetById=920273;ARGS:json.data,\
-    ctl:ruleRemoveTargetById=932236;ARGS:json.data,\
-    ctl:ruleRemoveTargetById=942200;ARGS:json.data,\
-    ctl:ruleRemoveTargetById=942260;ARGS:json.data,\
-    ctl:ruleRemoveTargetById=942340;ARGS:json.data,\
-    ctl:ruleRemoveTargetById=942370;ARGS:json.data,\
-    ctl:ruleRemoveTargetById=942430;ARGS:json.data,\
-    ctl:ruleRemoveTargetById=942431;ARGS:json.data,\
-    ctl:ruleRemoveTargetById=942432;ARGS:json.data"
+    ctl:ruleRemoveTargetById=920273;ARGS,\
+    ctl:ruleRemoveTargetById=932236;ARGS,\
+    ctl:ruleRemoveTargetById=942200;ARGS,\
+    ctl:ruleRemoveTargetById=942260;ARGS,\
+    ctl:ruleRemoveTargetById=942340;ARGS,\
+    ctl:ruleRemoveTargetById=942370;ARGS,\
+    ctl:ruleRemoveTargetById=942430;ARGS,\
+    ctl:ruleRemoveTargetById=942431;ARGS,\
+    ctl:ruleRemoveTargetById=942432;ARGS,\
```

Comment:
Why are you disabling these rules for all ARGS?

Reply:
Because, unfortunately, keys aren't hierarchical. For example, for the structure `{"one": "{"two": "three"}}`, ModSecurity will create the arg `one.two` with value `three` but no key `one` will exist. It's also not possible to match with a regular expression (at least it didn't work when I tried).

For a case like this, there are two options:
- exclude all `ARGS`
- list all possible keys

I wasn't sure that the object in the test contained all possible keys, so I opted for excluding `ARGS`.

Reply:
The data looks like this, I've just redacted the actual data while preserving the general structure.
```
{"data":"{\"id\":\"ZZ7gbdsfkK-BUODFAbiujsdf7-is\",\"type\":\"public-key\",\"rawId\":\"h7GEJH79Ds+ub86g8UGAS+Hs=\",\"response\":{\"authenticatorData\":\"3/zX+yiugsdf678567GYUSHADF/GQ+r657HKJGSFAD756==\",\"clientDataJSON\":\"iyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnugiyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnugiyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnugiyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnugiyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnug=\",\"signature\":\"yb8g869gBHJKAFSHJ+yYUUV7855687+Kq+ldfs86\",\"userHandle\":\"UHGASF=\"}}"}
```
So if I'm understanding this correctly, if there was a false positive within rawId the matched var would be ``json.data.rawId``? This doesn't line up with what's shown in my logs, the matched variable is ``json.data`` and not something else.
Maybe it's a difference between V2 and V3?

Reply:
> if there was a false positive within rawId the matched var would be `json.data.rawId`?

Yes. Can you show your logs?

Reply:
```
2024/03/03 23:34:50 [info] 3826458#3826458: *53 ModSecurity: Warning. Matched "Operator `Rx' with parameter `((?:[~!@#\$%\^&\*\(\)\-\+=\{\}\[\]\|:;\"'\xc2\xb4\xe2\x80\x99\xe2\x80\x98`<>][^~!@#\$%\^&\*\(\)\-\+=\{\}\[\]\|:;\"'\xc2\xb4\xe2\x80\x99\xe2\x80\x98`<>]*?){12})' against variable `ARGS:json.data' (Value: `{"id":"ZZ7gbdsfkK-BUODFAbiujsdf7-is","type":"public-key","rawId":"h7GEJH79Ds+ub86g8UGAS (458 characters omitted)' ) [file "/etc/nginx/modsecurity/coreruleset/rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf"] [line "1296"] [id "942430"] [rev ""] [msg "Restricted SQL Character Anomaly Detection (args): # of special characters exceeded (12)"] [data "Matched Data: {\x22id\x22:\x22ZZ7gbdsfkK-BUODFAbiujsdf7-is\x22,\x22type\x22:\x22 found within ARGS:json.data: {\x22id\x22:\x2ZZ7gbdsfkK-BUODFAbiujsdf7-is\x22,\x22type\x22:\x22public-key\x22,\x22rawId\x22:\x22Zc6BjnF9TU dH7H (462 characters omitted)"] [severity "4"] [ver "OWASP_CRS/4.0.0"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-sqli"] [tag "OWASP_CRS"] [tag "capec/1000/152/248/66"] [tag "PCI/6.5.2"] [tag "paranoia-level/2"] [hostname "0.0.0.0"] [uri "/login/webauthn/finish"] [unique_id "170946929077.753528"] [ref "o0,60o0,60v10,558t:urlDecodeUni"], client: 0.0.0.0, server: cloud.example.com, request: "POST /login/webauthn/finish HTTP/2.0", host: "cloud.example.com"
```

Reply:
Can you show full audit log for that request?

Reply:
I'm not sure what your looking for, libModSecurity3 doesn't log the request body. Everything else is just different rules that are triggered.
```
---7JWi0PR3---A--
[03/Mar/2024:23:34:50 +1100] 170946929077.753528 0.0.0.0 54272 0.0.0.0 443
---7JWi0PR3---B--
POST /login/webauthn/finish HTTP/2.0
sec-gpc: 1
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36
sec-fetch-site: same-origin
sec-ch-ua-mobile: ?0
origin: https://cloud.example.com
content-type: application/json
requesttoken: 42hg+fdshgfdh=:54/dfgh=
sec-ch-ua-platform: "Windows"
accept: application/json, text/plain, */*
sec-ch-ua: "Chromium";v="122", "Not(A:Brand";v="24", "Brave";v="122"
accept-language: en-US,en;q=0.8
content-length: 599
host: cloud.example.com
sec-fetch-mode: cors
sec-fetch-dest: empty
accept-encoding: gzip, deflate, br
```

Do you want all of the triggered rules as well?

Reply:
> libModSecurity3 doesn't log the request body

Are you sure? This is not possible to debug without request body.

Reply:
Yes I'm sure, it does log the response body though.
But isn't what I gave earlier enough? that's the payload sent by the client (according to chrome's inspect element)
Otherwise, I could try finding the full request with wireshark.

Reply:
I just noticed that the data is really strange - it looks like that Nextcloud is sending JSON data as a string inside a JSON:
`{"data":"{\"id\":\"ZZ7gbdsfkK-BUODFAbiujsdf7-is\",\"type\":\"public-key\",\"rawId\":\"h7GEJH79Ds+ub86g8UGAS+Hs=\",\"response\":{\"authenticatorData\":\"3/zX+yiugsdf678567GYUSHADF/GQ+r657HKJGSFAD756==\",\"clientDataJSON\":\"iyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnugiyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnugiyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnugiyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnugiyuhbsbbnl68567856876tghyiYUVUVASFDYUV79bnug=\",\"signature\":\"yb8g869gBHJKAFSHJ+yYUUV7855687+Kq+ldfs86\",\"userHandle\":\"UHGASF=\"}}"}`

Notice that everything inside value for "data" element is between `"` and all other `"` insde the data are escaped.

Reply:
Yes, that's what I'm seeing. Because the tests now use proper json structure, the output looks different from yours. The only explanation is what @azurit suggests.

Reply:
I'll update the tests and the rules.

Reply:
@theseion Cool! I also suggest replacing all 942* exclusions (there are too many, it's not worth it) with this:
`ctl:ruleRemoveTargetByTag=attack-sqli;ARGS:json.data`